File: code-challenges/array-binary-search/BinarySearch.java
import java.util.ArrayList;

public class BinarySearch {
  public static void main(String[] args) {
    Integer[][] arrays = new Integer[10][];
    for (int i = 0; i < arrays.length; i++) {
      Integer[] array = new Integer[i];
      for (int j = 0; j < i; j++) {
        array[j] = j + 1;
      }
      arrays[i] = array;
    }

    for (Integer[] array : arrays) {
      System.out.println("When working with "+arrToString(array)+":");
      for (int i = 0; i <= array.length + 1; i++) {
        int result = binarySearch(array, i);
        int expected = i == 0 || i == array.length + 1 ? - 1 : i - 1;
        System.out.println("Result (" + result + ") should be " + expected + " for element " + i);
        assert result == expected;
      }
    }
    System.out.println("All tests pass, yay!");
    //TODO test binary search/
  }

  public static <T> String arrToString(T[] arr) {
    StringBuilder string = new StringBuilder(2 + (arr.length * 3));
    string.append('[');
    for (int i = 0; i < arr.length; i++) {
      string.append(arr[i]);
      if (i < arr.length - 1) {
        string.append(", ");
      }
    }
    string.append(']');
    return string.toString();
  }

  public static <T extends Number> int binarySearch(T[] arr, T key) {
    int start = 0;
    int end = arr.length;
    int index = 0;
    double keyDouble = key.doubleValue(); //So we don't keep calling this.
    while (start < end) {
      index = ((end - start) / 2) + start;
      if (arr[index].equals(key)) {
        break;
      }
      if (arr[index].doubleValue() > keyDouble) {
        end = index - 1;
      } else {
        start = index + 1;
      }
    }
    index = ((end - start) / 2) + start;
    if (index >= arr.length || arr[index] != key) {
      return -1;
    }
    return index;
  } 
}

File: code-challenges/array-binary-search/binary-search.py
def binarySearch(arr, key):
  start = 0
  end = len(arr)
  index = 0
  while start < end:
    index = ((end - start) // 2) + start
    if arr[index] is key:
      break
    if arr[index] > key:
      end = index - 1
    else:
      start = index + 1
  index = ((end - start) // 2) + start
  if index not in range(len(arr)) or arr[index] is not key:
    return -1
  return index

if __name__ == '__main__':
  arrays = []
  for i in range(10):
    array = []
    for j in range(i):
      array.append(j + 1)
    arrays.append(array)

  for array in arrays:
    print(f'When working with {array} as arr:')
    for i in range(0, len(array) + 2):
      result = binarySearch(array, i)
      expected = -1 if i == 0 or i == len(array) + 1 else i - 1
      print(f'Result ({result}) should be {expected} for element {i}')
      assert result is expected
  print('All tests pass, yay!')
